internal/service/handlers/activate_balance.go

```diff
@@ -0,0 +1,147 @@
+package handlers
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/rarimo/geo-auth-svc/pkg/auth"
+	"github.com/rarimo/geo-points-svc/internal/data"
+	"github.com/rarimo/geo-points-svc/internal/data/evtypes"
+	"github.com/rarimo/geo-points-svc/internal/data/evtypes/models"
+	"github.com/rarimo/geo-points-svc/internal/service/requests"
+	"gitlab.com/distributed_lab/ape"
+	"gitlab.com/distributed_lab/ape/problems"
+)
+
+func ActivateBalance(w http.ResponseWriter, r *http.Request) {
+	req, err := requests.NewActivateBalance(r)
+	if err != nil {
+		ape.RenderErr(w, problems.BadRequest(err)...)
+		return
+	}
+
+	nullifier := req.Data.ID
+	referralCode := req.Data.Attributes.ReferredBy
+
+	if !auth.Authenticates(UserClaims(r), auth.UserGrant(nullifier)) {
+		ape.RenderErr(w, problems.Unauthorized())
+		return
+	}
+
+	balance, err := BalancesQ(r).FilterByNullifier(nullifier).Get()
+	if err != nil {
+		Log(r).WithError(err).Error("Failed to get balance by nullifier")
+		ape.RenderErr(w, problems.InternalError())
+		return
+	}
+
+	if balance == nil {
+		Log(r).Debugf("Balance %s not exist", nullifier)
+		ape.RenderErr(w, problems.NotFound())
+		return
+	}
+
+	if balance.ReferredBy != nil {
+		Log(r).Infof("Balance already activated with code '%s'", *balance.ReferredBy)
+		ape.RenderErr(w, problems.Conflict())
+		return
+	}
+
+	referral, err := ReferralsQ(r).FilterInactive().Get(req.Data.Attributes.ReferredBy)
+	if err != nil {
+		Log(r).WithError(err).Error("Failed to get referral by ID")
+		ape.RenderErr(w, problems.InternalError())
+		return
+	}
+	if referral == nil {
+		Log(r).Debugf("Referral code '%s' not found", referralCode)
+		ape.RenderErr(w, problems.NotFound())
+		return
+	}
+
+	refBalance, err := BalancesQ(r).FilterByNullifier(referral.Nullifier).Get()
+	if err != nil || refBalance == nil { // must exist due to FK constraint
+		Log(r).WithError(err).Error("Failed to get referrer balance by nullifier")
+		ape.RenderErr(w, problems.InternalError())
+		return
+	}
+
+	err = EventsQ(r).Transaction(func() error {
+		level, err := doLvlUpAndReferralsUpdate(Levels(r), ReferralsQ(r), *balance, 0)
+		if err != nil {
+			return fmt.Errorf("failed to do lvlup and referrals update: %w", err)
+		}
+
+		err = BalancesQ(r).FilterByNullifier(balance.Nullifier).Update(map[string]any{
+			data.ColReferredBy: referralCode,
+			data.ColLevel:      level,
+		})
+		if err != nil {
+			return fmt.Errorf("update balance amount and level: %w", err)
+		}
+
+		if refBalance.ReferredBy != nil {
```

Comment:
A bit confusing. Add comment that we add referred_by when it's not genesis balance